Create Window_and_Application_API_Spec.md

marb2000 opened the pull request:
Adding draft of Window and Application API Specs to support Win32

Felix-Dev commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,618 @@
+```
+
+#### Remarks:
+- **Win32**: The image is set as an Icon of the Window. 
````
What happens when 'null' is specified for the icon? Do Win32 apps fall back to the WPF behavior which is displaying a default application icon or does it use the WinForms & UWP behavior and displays no icon?

mdtauk replied:
I would prefer there to be no Icon by default, which would mimic UWP behaviour

marb2000 replied:
This is a great suggestion. No showing up Icon unless you specify one. The TaskBar and Alt+Tab icon could be the application icon.

marb2000 replied:
Closing this conversation given Icon property is no longer a plan for 2020. We will discuss about it in the next update.

Felix-Dev commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,618 @@
+
+#### Remarks
+
+- **Win32**: This method do nothing when running in Win32.
```
There has been some thinking in WinUI to add UWP's titlebar customization functionality (i.e. CoreApplicationViewTitleBar.ExtendViewIntoTitleBar) to WinUI Desktop apps too. Seeing as Window.SetTitleBar is not supported on Win32 apps, does that mean that WinUI Desktop apps will still have to re-draw the entire titlebar and re-implement titlebar (drag, minimize, maximize, close,...) logic if they want to provide a custom titlebar?

marb2000 replied:
We are introducing an API in the XAML layer to extend the content into the Title bar. Resolving this conversation.

dotMorten commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,618 @@
+}
+```
+#### Remarks
+VisibilityChangedEventArgs includes a Boolean Visible property that you should check to determine the visibility state when the event was fired.
````
I'd like to see this entire visibility api moved to UIElement instead. See issue logged here with justifications:
https://github.com/microsoft/microsoft-ui-xaml/issues/674

MikeHillberg replied:
Obvious overlap, but these look like two separate issues. I agree we should have a UIElement.IsVisible property like WPF's, but the Window.Visible property already exists, and Window isn't a UIElement. So we can have both Window.Visible and UIElement.IsVisible.

dotMorten commented:
You `Activate` a Window to view it, but you `Close` it to remove it. That seems rather inconsistent. Should `Close` be `Deactiveate` or should `Activate` be `Open` ?

Also it's not clear to me whether `IsVisible` is true or false when the window is minimized. Should it be `IsOpen` is `IsActivated` ?

dotMorten commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,618 @@
+    - Exception Message: TBD. Something like "Only one Window per thread is allowed in UWP app model".
+
+### Window.Bounds property
+Gets a Rect value containing the height and width of the application window in logical pixels.
```
A `Rect` also provides location.  Is this the location on the current screen? Across all screens? Which screen is it on?
If it truly only provides height and width, this should be a `Size` type

MikeHillberg replied:
This is an existing API and this is the [existing text]([Window.Bounds](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Window.Bounds)), but agreed it's wrong. 
* In UWP this should be simply returning the [CoreWindow.Bounds](https://docs.microsoft.com/uwp/api/Windows.UI.Core.CoreWindow.Bounds).
* On Desktop this returns the GetClientRect adjusted for rasterization scale.

dotMorten commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,618 @@
+- **WindowEventArgs.Handled**: Gets or sets whether the event was handled. **true** if the event has been handled by the appropriate delegate; **false** if it has not.
+    ```CS
+    public bool Handled { get; set; }
+    ```
````
What's the point of handling this? Will that prevent it from resizing?

dotMorten commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,618 @@
+- **WindowActivatedEventArgs.Handled**: Gets or sets whether the window activation changed event was handled. **true** if the event has been handled by the appropriate delegate; **false** if it has not.
+    ```CS
+    public bool Handled { get; set; }
+    ```
````
What's the point of handling this? Will that prevent it from getting activated? If I don't handle it, what will it bubble up to?

marb2000 replied:
We kept it for backwards compatibility with the existing APIs. Honestly, I didn't find a scenario that justify this, but perhaps my scope (from XAML) is narrower than the scope of UWP windowing creators.

dotMorten commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,618 @@
+- VisibilityChangedEventArgs.Handled: Gets or sets whether the window activation changed event was handled. **true** if the event has been handled by the appropriate delegate; **false** if it has not.
+    ```CS
+    public bool Handled { get; set; };
+    ```
````
What's the point of handling this? Will that prevent it from becoming visible or hidden?

marb2000 replied:
Same comment as above

dotMorten commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,618 @@
+
+- WindowSizeChangedEventArgs.Handled: Gets or sets whether the  event was handled. **true** if the event has been handled by the appropriate delegate; **false** if it has not.
+    ```CS
+    public bool Handled { get; set; };
````
What's the point of handling this? Will that prevent it from resizing?

marb2000 replied:
Some comment as above

dotMorten commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,618 @@
+<Application />
+ ```
+
+> **NOTE**: It is not a goal of this doc to review all the members of this class. Just that ones that requires changes in Win32. You can find more inforation about Application class [here](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.application).
````
I wish this was on top, because I just did that :)

marb2000 replied:
We change the scope of this spec, and only cover the updates, so this comment is outdated. Resolving it.

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,618 @@
+  }
+}
+```
+##### Multiple Widndows Sample
````
>Widndows [](start = 15, length = 8)

Spelling

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,618 @@
+        w.Activate();
+        w.Closed += (sender1, e1) => w.Dispatcher.InvokeShutdown();
+
+        System.Windows.Threading.Dispatcher.Run();
```
>System [](start = 8, length = 6)

Which Dispatcher did you intend here?

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,618 @@
+```CS
+private void Button_Click(object sender, RoutedEventArgs e)
+{
+    Thread thread = new Thread(() =>
````
>Thread [](start = 4, length = 6)

Does it make sense to have a Xaml worker function / class that makes this more maintainable, abstracts Win32 and UWP, and enables us to service? May we follow-up?

marb2000 replied:
This is a good point, but out of scope for 2020. Adding a note to capture this in the backlog.

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,618 @@
+Gets a Rect value containing the height and width of the application window in logical pixels.
+
+```CS
+public Rect Bounds { get; }
````
>Bounds [](start = 12, length = 6)

A "Rect" is particularly problematic. Can we change this to ClientSize or something? We'll want to distinguish from FrameSize, which is in a different coordinate space and affected by borders, etc.

MikeHillberg replied:
This is an existing API. But see the [other comment](https://github.com/microsoft/microsoft-ui-xaml-specs/pull/71/files/9c72d8dde073924516b125e1ddaab1de36c61bc4#r393373967) about this property.

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,618 @@
+public class Window : IClosable
+```
+
+### **Window management, appearance, and behaviors**
````
>Window management, appearance, and behaviors [](start = 6, length = 44)

Please take a look at AppWindow and Forms for examples of basic knobs to customize the Window (frame style, show title bar, etc.):
[https://docs.microsoft.com/en-us/uwp/api/windows.ui.windowmanagement.appwindow](https://docs.microsoft.com/en-us/uwp/api/windows.ui.windowmanagement.appwindow)
[https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.form?view=netframework-4.8](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.form?view=netframework-4.8)

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,618 @@
+</Window>
+```
+```CS
+public partial class MainWindow:Window
````
>MainWindow:Window [](start = 21, length = 17)

Please write "MainWindow : Window", with the spaces.

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,618 @@
+
+#### Remarks:
+- **Win32**: The image is set as an Icon of the Window. 
+- **UWP**: Do nothing. The property value is ignored.
```
Does it make sense to have a more unified way to specify the icon between Win32 and UWP? I think it comes from the manifest for UWP.

mdtauk replied:
With Live Tiles displaying Icons, it probably makes sense to ensure a single ico file can be used to cover all the uses, without having to duplicate.

marb2000 replied:
Icon property was moved out of 2020. We will discuss this when we add this feature in the future.

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,618 @@
+VisibilityChangedEventArgs includes a Boolean Visible property that you should check to determine the visibility state when the event was fired.
+
+### Window.SetTitleBar(UIElement) Method
+Set the custom XAML content that should act as the title bar.
```
This is out of scope for Nov 2020 for Win32 apps.

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,618 @@
+
+```CS
+public static Window Current { get; }
+```
````
IMPORTANT: Absolutely do not do this! There is no "current" window when we support multiple windows on a single thread. Trying to store this and update it "depending on the operation" is a very problematic issue that does not have any good answers. It's been well researched for the last decade as part of the ApplicationView / CoreWindow problems with GetForCurrentView.

MikeHillberg replied:
This is an existing API. The purpose of calling it out in this spec is to add the below documentation which explains that it returns null in Desktop apps for that reason.

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,618 @@
+#### Remarks
+- **UWP**: UWP app typically use a single Window instance, and do not open and close additional Window instances. Additionally, you will not typically provide custom UI to enable users to close the app window.
+
+### Window.Closed event
```
>Closed [](start = 11, length = 6)

Do you want to have typical functionality here:
1. Allow app to get a deferral and start saving state.
2. Potentially override the "closing" behavior.

I'd recommend talking with Roberth and Francis on real-world scenarios here.

marb2000 replied:
Thanks for the piece of advice!  We will talk to them. Resolving this conversation,

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,618 @@
+These properties expose these underneath objects to enable advanced (or basic but not supported yet) scenarios, so developers can get the CoreWindow or HWnd objects to apply setters, listening events, etc.
+
+### Window.CoreWindow property: 
+
```
Xaml's "Window" is going to become huge if it becomes a dumping ground for properties that will rarely be used. Can we move these to Hosting as a static member instead?

static class UwpWindowHosting
{
    static ApplicationView GetView(Window window);
    static CoreWindow GetCoreWindow(Window window);
}

Or, just have them call CoreWindow.GetForCurrentThread() and ApplicationView.GetForCurrentView() like everyone has done since Win 8.x.

dotMorten replied:
@jeffstall didn't you just say in the above comment
> There is no "current" window when we support multiple windows on a single thread

Is so, how would CoreWindow.GetForCurrentThread() work?

MikeHillberg replied:
I think his point is that rather than having instance members on the Window class, we should have static members on helper classes that take a Window instance as their first parameter.

We've done this for dumping grounds before (like [Typography](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Documents.Typography)), but in general it serves to make APIs undiscoverable. And I don't see Window becoming a dumping ground (WPF's [Window](https://docs.microsoft.com/dotnet/api/System.Windows.Window) is an existing precedent to look at).

marb2000 replied:
XAML's Window will live with Current for backwards compatibility. Perhaps we can deprecate it in the future, but not now.

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,618 @@
+
+```CS
+public CoreDispatcher Dispatcher { get; }
+```
````
Change to the DispatcherQueue, since CoreDispatcher isn't that interesting.

MikeHillberg replied:
It's an existing property. But DispatcherQueue is being added

dotMorten commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -405,12 +405,79 @@ Provides event information when an app is launched.
+# Appendix
+
+Changes in this spec from Preview2:
+* Rename LaunchActivatedEventArgs to XamlLaunchActivatedEventArgs
```
Isn't that already implied from the namespace? IMHO I find it a bit weird to prefix this with `Xaml` for an application that launched.

MikeHillberg replied:
That's to try and avoid confusion with the existing type named [LaunchActivatedEventArgs](https://docs.microsoft.com/uwp/api/Windows.ApplicationModel.Activation.LaunchActivatedEventArgs) (in the Windows.ApplicationModel.Activation namespace). This Xaml type is a wrapper around that.

Another alternative could be to simply give this a different name; "launch activated" is a bit confusing in this context. For example `ApplicationActivatedEventArgs`.

marb2000 replied:
It seems that the preferred was XamlLaunchActivatedEventArgs. Resolving the conversation

marb2000 commented:
Closing this PR given we already moved and updated the Spec [here](
https://github.com/microsoft/microsoft-ui-xaml-specs/blob/Win32/active/Win32/Window_and_Application_API_Spec.md)

marb2000 commented:
Sorry, I mistake.

dotMorten commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -118,10 +115,41 @@ thread.Start();
+
+    //Assuming TitleBar never is null
+    m_window.ExtendContentIntoTitleBar = true;
+    m_window.SetTitleBar(m_window.CustomTitleBar);
```
Why not just `m_window.TitleBar = m_window.CustomTitleBar;` ?

marb2000 replied:
This is good question. I just kept it for compatibility with UWP because it's an existing API, but I don't know the reasons of why we did in this way long time ago. @MikeHillberg any idea?

MikeHillberg replied:
I don't recall either. I do remember we talked about adding a Get method. We in general have a guideline that property setters and getters should be trivial, and if they do any real work should be methods (though we've probably violated that in a few places to be friendlier to markup). Maybe that kicked in here.

MikeHillberg replied:
Leaving this as a method since it's an existing API.

dotMorten commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -433,6 +458,7 @@ Microsoft.UI.Xaml namespace
         String Title;
         ImageSource Icon;
         Microsoft.System.DispatcherQueue DispatcherQueue{ get; };
+	Boolean ExtendContentIntoTitleBar { get; set;}	
```
nit: indent

marb2000 replied:
Fixed.

mdtauk commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -523,6 +547,28 @@ In a UWP app, this is equivalent to LaunchActivatedEventArgs.Arguments.
 
 # API Details
 
+?? Need to cover resources:
```
You could use ` ```XAML ` to add context colour to this code.

Also when you say Need to cover - does that mean you will still be able to set these values in Xaml for both UWP and Win32 **Window**s that are created in WinUI 3?

MikeHillberg replied:
I'm removing this section, we're not trying yet to make this part of the API.

stevenbrix commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,654 @@
+
+This method is typically used with the Window's `ExtendContentIntoTitleBar` property 
+set to true in order to hide the default system title bar.
+However, even when the default system title bar is not hidden, SetTitleBar can be used to make 
```
what is the user scenario for making any part of the app behave like the title bar? Since we want a consistent look and feel across all Windows apps, it would seem like that in the 97% use case, we don't want apps to differentiate from the default behavior. I'm sure there is a valid scenario where this is desired, and it would be helpful for the uninformed to understand when that is.

mdtauk replied:
> what is the user scenario for making any part of the app behave like the title bar? Since we want a consistent look and feel across all Windows apps, it would seem like that in the 97% use case, we don't want apps to differentiate from the default behavior. I'm sure there is a valid scenario where this is desired, and it would be helpful for the uninformed to understand when that is.
>
Pretty much any modern UWP app will extend into the Titlebar area, and will place a XAML element there so you can still grab the top of the window to move the window around.

![image](https://user-images.githubusercontent.com/7389110/101952952-673adc80-3bf1-11eb-9c49-b115f351b10b.png)
_Settings_

![image](https://user-images.githubusercontent.com/7389110/101952995-7cb00680-3bf1-11eb-8a68-70a95b5e954e.png)
_Your Phone_

![image](https://user-images.githubusercontent.com/7389110/101953037-8c2f4f80-3bf1-11eb-809a-433ce18aed36.png)
_Calculator_

![image](https://user-images.githubusercontent.com/7389110/101953103-ab2de180-3bf1-11eb-884f-d0fde346ffc7.png)
_Heutro_

![image](https://user-images.githubusercontent.com/7389110/101953138-bf71de80-3bf1-11eb-985e-d1903aa9fa1a.png)
_Windows Store_

![image](https://user-images.githubusercontent.com/7389110/101953198-dadce980-3bf1-11eb-96b3-66a1d56bd15e.png)
_Alarms and Clock_

![image](https://user-images.githubusercontent.com/7389110/101953285-ffd15c80-3bf1-11eb-9bc6-1b5f8a7fa116.png)
_Maps_

![image](https://user-images.githubusercontent.com/7389110/101953343-1a0b3a80-3bf2-11eb-936c-936494f21280.png)
_Messenger (React Native)_

stevenbrix replied:
Sorry, I must not have explained myself well. I'm not suggesting that extending into the title bar doesn't have value. 

But the spec, and implementation, allows me to do something like this:

![image](https://user-images.githubusercontent.com/9649518/101961020-a94a2b00-3bbd-11eb-8c33-34056d3a1745.png)

Where the "title bar" is now in the center of my app. And, unlike the spec, I can no longer interact with the top of the Window and drag it. It doesn't matter if I set `CoreApplicationViewTitleBar.ExtendViewIntoTitleBar` or not. As soon as I call `Window.SetTitleBar` the top of the window is no longer draggable

mdtauk replied:
Apps which Extend into the titlebar, will often put items into the titlebar area, like the "What's new" button in the Your PC app.

The SetTitleBar API is used to allow the hit target of the panel placed where the Titlebar would be, to be used for dragging, snapping, and double clicking to maximise a window.  Without it, anything you place in that area would eat the taps and clicks, and it would not be possible to move the windows.

``
This method lets you specify a XAML element that interacts with the system as if it’s the app window's title bar. When you use this method, you typically set the CoreApplicationViewTitleBar.ExtendViewIntoTitleBar property to true in order to hide the default system title bar. However, even when the default system title bar is not hidden, this can be used to make additional regions in your app behave like the title bar.
``

https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.window.settitlebar?view=winrt-19041

MikeHillberg replied:
I don't understand/recall the motivation for the existing API to allow you to make an element in the middle of the Window be the title bar (for purposes of drag etc). A Window.TitleBar property sounds more useful to me. For the moment though we're just enabling the existing API on Desktop.

mdtauk commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,662 @@
+}
+```
+
+Title bars should be have background and foreground colors based on
````
"Title bars should have background"... the **be** is not needed

mdtauk commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,662 @@
+
+## Window.ExtendsContentIntoTitleBar
+
+Gets or sets a value that specifies whether the default window title bar should be removed, creating more space for content.
```
Rather than saying it makes more space, might it not be better and more accurate to say something like **it allows the App's UI to extend to the top of the window's frame, where the titlebar would usually be.**

mdtauk commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,662 @@
+However, even when the default system title bar is not hidden, SetTitleBar can be used to make 
+additional regions in your app behave like the title bar.
+
+The following shows a Window with a no built-in title bar. 
```
"The following shows a Window with ~a~ no built-in title bar." remove **a**

mdtauk commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,662 @@
+
+The following shows a Window with a no built-in title bar. 
+Instead it uses the whole window as its content area. To enable features such as window dragging, 
+it designates a TextBlock as the title bar.
```
Do you want the example to use the TextBlock as the titlebar, or the <Grid> containing the TextBlock?  If so, the Grid should also use a background color of **Transparent** for hit testing.

mdtauk commented:
Will there be support for WinUI 3 Windows which can hide the Minimise and Maximise buttons, much like the modal dialogs seen commonly in win32 apps - as well as MessageBoxes etc?  I know these will probably be ignored by UWP app windows.

![win32 dialog window](https://user-images.githubusercontent.com/7389110/104252083-1248eb00-5469-11eb-8ee2-8216795f63cd.png)
_Close only_

![Windows 10X titlebar](https://user-images.githubusercontent.com/7389110/104252239-78357280-5469-11eb-806c-e3c4d4ce26ec.png)
_10X titlebar with no Maximise/Restore button_

jonwis commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,655 @@
+In a UWP app you must create a new 
+[CoreApplicationView](https://docs.microsoft.com/uwp/api/Windows.ApplicationModel.Core.CoreApplicationView),
+which will automatically create a new thread and a Window on it. 
+See the 'Create new Window in UWP application' example.
```
What if I `new` a Window while on the Dispatcher for a CoreWindow? Could that create me a new [AppWindow](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.WindowManagement.AppWindow?view=winrt-19041) instead?

jonwis commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,655 @@
+object for the Window when called from a UWP app. Returns null when called from a Desktop app.
+
+```CS
+public CoreDispatcher Dispatcher { get; }
````
So we end up with both `Dispatcher` and `DispatcherQueue` as aliases of each other? Should this property be insta-deprecated to help developers move to the preferred one instead?

jonwis commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,655 @@
+
+### Remarks
+
+The following events and virtual methods are not invoked when running in a Desktop app:
```
Maybe link over to the Project Reunion issue thread on Win32-facing activation events so folks know they're not out of luck?

marb2000 reviewed (commented):
I don't get it. This is a C# sample, and in C# WindowHandle returns IntPtr. HWND is not even a type. Why this change?

oldnewthing commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,655 @@
+with non-pointer input devices such as an Xbox gamepad or remote control. (All nested elements inherit 
+this behavior.)
+
+> Note: this property is ignored in a Desktop app
```
So does this mean that Desktop apps always support mouse mode? Or Desktop apps never support mouse mode? Please clarify by saying what the Desktop behavior is:

> Note: This property is ignored in a Desktop app. For Desktop apps, pointer mode is always [ Never | WhenEngaged | WhenFocused | something else ].

oldnewthing commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,655 @@
+### Remarks
+- In a UWP app, this method is called once for the creation of the main window, once for each hosted window, 
+and once each time CreateNewView is called. 
+- In a Desktop app, this method is called every time a Window object is created.
```
I assume this method is called from the thread that created the window.

oldnewthing commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,655 @@
+```
+
+### Remarks
+- In a UWP app, this method is called once for the creation of the main window, once for each hosted window, 
````
For a UWP app, does this mean that OnWindowCreated is called before OnLaunched? This would be a gotcha case for many apps, since they typically do the bulk of their initialization in OnLaunched. May want to add a Note for this case.

oldnewthing commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,655 @@
+**Handled**
+
+Gets or sets whether the window activation changed event was handled.
+ **true** if the event has been handled by the appropriate delegate; **false** if it has not.
```
This is an "-ed" event, so the activation has already occurred. The Handled would presumably merely prevent other event handlers from seeing the event? That seems like a weird thing to do. What's the use case for handling the event?

oldnewthing commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,655 @@
+
+**Handled**
+
+Gets or sets whether the window activation changed event was handled. **true** 
```
copy/pasta. (Also, same question about what "Handled" means here.)

oldnewthing commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,655 @@
+
+**Visible**
+
+Gets whether the window is visible or not. **true** if the window is visible; otherwise is **false**.
```
"is" or "was"? This is an "-ed" event, so the property presumably refers to whether the window was made visible or not, rather than referring to the window's current visibility state (which you can obtain from the Window.Visible property.)

oldnewthing commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,655 @@
+**Arguments**
+
+Gets the arguments that are passed to the app during its launch. 
+In a UWP app, this is equivalent to LaunchActivatedEventArgs.Arguments.
```
I assume that for a Desktop app, the Arguments is the raw command line that comes after the program name (leading spaces preserved)? I.e., we don't try to parse it in any way with CommandLineToArgvW or anything like that?

oldnewthing commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,655 @@
+
+    // These APIs match the system Xaml versions, except the event args are now from the Xaml namespace
+    event Windows.Foundation.TypedEventHandler<Object,Microsoft.UI.Xaml.XamlWindowActivatedEventArgs> Activated;
+    event Windows.Foundation.TypedEventHandler<Object,Microsoft.UI.Xaml.XamlWindowClosedEventArgs> Closed;
```
Missing from spec. Did we intend to keep this? Or did we forget to delete it from the IDL?

oldnewthing commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,655 @@
+
+```cs
+[webhosthidden]
+runtimeclass XamlWindowClosedEventArgs
````
Missing from spec. Did we intend to keep this? Or did we forget to delete it from the IDL?

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,655 @@
+with this Window. In a Desktop app this is always null.
+
+```CS
+public CoreWindow CoreWindow { get; }
````
There's no reason to expose CoreWindow and CoreDispatcher. If UWP apps really need these, just use the existing static methods on those types. (i.e. CoreWindow.GetForCurrentThread)

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,655 @@
+    Window window = new Window();
+    window.Content = new TextBlock() { Text = "Hello" };
+    window.Activate();
+    window.Closed += (_, __) => w.DispatcherQueue.InvokeShutdown();
```
Sample feedback. Please use real variable names, not "_".

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,655 @@
+the Window's underlying HWND.
+
+```CS
+protected override void OnLaunched(Microsoft.UI.Xaml.XamlLaunchActivatedEventArgs e)
````
I don't understand this sample. Is it in C#, which wouldn't have the IWindowNative interface definition, or in C++, which wouldn't have this language syntax.

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
```diff
@@ -0,0 +1,655 @@
+
+> Spec note: Application is not new to this spec, what is new are the following remarks
+
+### Remarks
```
Should we do some refactoring here, since it gets confusing if we have a lot of things that never work for Win32, and this is becoming a front and center scenario?

jeffstall commented on active/Win32/Window_and_Application_API_Spec.md:
````diff
@@ -0,0 +1,655 @@
+```cs
+[webhosthidden]
+[contentproperty("Content")]
+unsealed runtimeclass Window
````
Important to specify event order. Does VisibilityChanged get fired before or after SizeChanged? Does VisibilityChanged get fired when minimized?

sylveon replied:
If visibility maps to ShowWindow(SW_HIDE/SW_SHOW) and IsWindowVisible, I don't think it should fire, since a minimized app is still "visible".
